add skipNode() function in TreeNode

amymeng5277 opened the pull request.

amymeng5277 committed:
Merge pull request #8 from foam-framework/master

Update

amymeng5277 committed:
1.Fixed skip() function of MDAO. 2. Add SkipNode() function in TreeNode

amymeng5277 committed:
Merge pull request #9 from foam-framework/master

sync from master

amymeng5277 committed:
Merge branch 'fixTreeindex' of https://github.com/amymeng5277/foam2 into fixTreeindex

amymeng5277 committed:
1. add skipNode() function in TreeNode

kgrgreer commented on src/foam/dao/index/TreeNode.java:
```diff
@@ -357,19 +359,58 @@ public TreeNode lte(TreeNode s, Object key, PropertyInfo prop) {
+  }
+
+  public void select(TreeNode currentNode, Sink sink, long skip, long limit, Comparator order, Predicate predicate) {
+    if(skip>currentNode.size || limit == 0)
```
<= 0

amymeng5277 committed:
Merge pull request #10 from foam-framework/master

sync from master

amymeng5277 committed:
1.Fixed space in TreeNode 2. Add decorateSink in MDAO, for other select plan still working on.

kgrgreer commented on src/foam/dao/MDAO.java:
```diff
@@ -63,6 +63,7 @@ public FObject find_(X x, Object o) {
   }
 
   public Sink select_(X x, Sink sink, long skip, long limit, Comparator order, Predicate predicate) {
+    sink = decorateSink_(sink,skip,limit,order,predicate);
```
You shouldn't do this here. Inside the plan you should add whatever Sink decorators are required because you couldn't handle them efficiently, but it's too early to tell at this point which ones you can handle efficiently.

kgrgreer commented on src/foam/dao/index/TreeNode.java:
```diff
@@ -357,19 +359,56 @@ public TreeNode lte(TreeNode s, Object key, PropertyInfo prop) {
+  }
+
+  public void select(TreeNode currentNode, Sink sink, long skip, long limit, Comparator order, Predicate predicate) {
+    if( skip >= currentNode.size || limit <= 0 ) return;
```
space between if and (

kgrgreer commented on src/foam/dao/index/TreeNode.java:
```diff
@@ -357,19 +359,56 @@ public TreeNode lte(TreeNode s, Object key, PropertyInfo prop) {
+
+  public void select(TreeNode currentNode, Sink sink, long skip, long limit, Comparator order, Predicate predicate) {
+    if( skip >= currentNode.size || limit <= 0 ) return;
+    if( skip == 0 && limit > currentNode.size ){
```
Format as: if ( skip == 0 && limit > currentNode.size ) { (two extra spaces)

kgrgreer commented on src/foam/dao/index/TreeNode.java:
```diff
@@ -357,19 +359,56 @@ public TreeNode lte(TreeNode s, Object key, PropertyInfo prop) {
+      selectAllHelper(currentNode,sink);
+      return;
+    }
+    if( skip != 0 ){
```
fix spacing

amymeng5277 committed:
Merge pull request #11 from foam-framework/master

sync from master

amymeng5277 committed:
Merge pull request #12 from foam-framework/master

sync from master

amymeng5277 committed:
Merge pull request #13 from foam-framework/master

sync

amymeng5277 committed:
Merge pull request #14 from foam-framework/master

sync from master

amymeng5277 committed:
1.Eq,Lte,Lt,Gt,Gte now worked well for index search.

amymeng5277 committed:
Merge pull request #15 from foam-framework/master

sync from master

amymeng5277 committed:
1.Implement EQ(),GTE(),GT(),LT(),LTS(), skip(),limit() such function for MDAO, such function now supported UnIndex search.
2. Add AltSelectPlan to remember predicate, state and plan.
3.Fixed SkipTreeNode() to support Ununique Index select.

kgrgreer commented on src/foam/dao/index/ScanPlan.java:
```diff
@@ -0,0 +1,32 @@
+
+  public long cost() { return cost_; }
+
+  public void setCost(long cost) { cost_ = cost;}
```
I would just pass this in the constructor.

kgrgreer commented on src/foam/dao/index/ScanPlan.java:
```diff
@@ -0,0 +1,32 @@
+  }
+
+  public void select(Object state, Sink sink, long skip, long limit, Comparator order, Predicate predicate) {
+    if(state == null)
```
fix spacing

kgrgreer commented on src/foam/dao/index/TreeIndex.java:
```diff
@@ -42,14 +38,31 @@ public Object bulkLoad(FObject[] a) {
     }
+    if ( predicate instanceof And ) {
+      int length = ((And) predicate).getArgs().length;
+      for(int i=0;i<length;i++){
```
spacing

kgrgreer commented on src/foam/dao/index/TreeIndex.java:
```diff
@@ -42,14 +38,31 @@ public Object bulkLoad(FObject[] a) {
+            ((Binary) predicate).getArg1().toString().equals(prop_.toString()) ) {
+
+          Binary b = new Binary() {};
+          ((And) predicate).getArgs()[i]= new True();
```
spacing

kgrgreer commented on src/foam/dao/index/TreeIndex.java:
```diff
@@ -84,8 +97,19 @@ public SelectPlan planSelect(Object state, Sink sink, long skip, long limit, Com
+    //expr = isExprMatch(predicate, In.class);
+    //if ( expr != null && Math.log(((TreeNode)state).size)/Math.log(2) * ((In) predicate) < ((TreeNode) state).size )
+    expr = isExprMatch(predicate, Eq.class);
+    if ( expr != null) {
```
spacing

kgrgreer commented on src/foam/dao/index/TreeIndex.java:
```diff
@@ -96,25 +120,46 @@ public SelectPlan planSelect(Object state, Sink sink, long skip, long limit, Com
+    if ( subTree == null ) {
+      cost = 0;
+    }
+    else {
```
} else {

kgrgreer commented on src/foam/dao/index/TreeIndex.java:
```diff
@@ -96,25 +120,46 @@ public SelectPlan planSelect(Object state, Sink sink, long skip, long limit, Com
+      cost = 0;
+    }
+    else {
+      cost = subTree instanceof TreeNode?((TreeNode)subTree).size : 1;
```
spacing

kgrgreer commented on src/foam/dao/index/TreeIndex.java:
```diff
@@ -96,25 +120,46 @@ public SelectPlan planSelect(Object state, Sink sink, long skip, long limit, Com
+    }
+    boolean sortRequired = false;
+    boolean reverseSort = false;
+    if ( order!=null ) {
```
spacing

kgrgreer commented on src/foam/dao/index/TreeIndex.java:
```diff
@@ -96,25 +120,46 @@ public SelectPlan planSelect(Object state, Sink sink, long skip, long limit, Com
+        reverseSort = true;
+      } else {
+        sortRequired = true;
+        if ( cost !=0 ) cost *= Math.log(cost) / Math.log(2);
```
spacing

kgrgreer commented on src/foam/dao/index/AltIndex.java:
```diff
@@ -89,12 +89,11 @@ public SelectPlan planSelect(Object state, Sink sink, long skip, long limit, Com
     }
 
-    return new AltSelectPlan(bestState, bestPlan);
+    return bestPlan;
```
Why don't you need the AltSelectPlan anymore?

kgrgreer commented on src/foam/dao/index/AltSelectPlan.java:
```diff
@@ -15,14 +15,16 @@
 public class AltSelectPlan implements SelectPlan {
   protected Object state_;
   protected SelectPlan bestPlan_ = new NotFoundPlan();
+  protected Predicate expr_;
```
What is this needed for?

kgrgreer commented on src/foam/dao/index/TreeIndex.java:
```diff
@@ -42,14 +37,31 @@ public Object bulkLoad(FObject[] a) {
+      int length = ((And) predicate).getArgs().length;
+      for (int i = 0; i < length; i++) {
+        Predicate arg = ((And) predicate).getArgs()[i];
+        if ( predicate instanceof Binary &&
```
These three lines are the same as line 39 above. Don't duplicate logic.

kgrgreer commented on src/foam/dao/index/TreeIndex.java:
```diff
@@ -42,14 +37,31 @@ public Object bulkLoad(FObject[] a) {
+            ((Binary) predicate).getArg1().toString().equals(prop_.toString()) ) {
+
+          Binary b = new Binary() {};
+          ((And) predicate).getArgs()[i]= new True();
```
spacing

kgrgreer commented on src/foam/dao/index/TreeIndex.java:
```diff
@@ -74,6 +86,35 @@ public FindPlan planFind(Object state, Object key) {
     return new TreeLookupFindPlan(prop_, (state != null ? ((TreeNode) state).size : 0) );
   }
 
+  public SelectPlan SelectPlanHelper(Object state, Sink sink, long skip, long limit, Comparator order, Predicate predicate) {
```
Helper is a generic name. If you find a more meaningful name, that would be better. A convention is also to just call it select_() if it's an internal helper for select(). Also, method names typically start with lower-case characters.

kgrgreer commented on src/foam/dao/index/TreeIndex.java:
```diff
@@ -74,6 +86,35 @@ public FindPlan planFind(Object state, Object key) {
+    boolean sortRequired = false;
+    boolean reverseSort = false;
+    if ( order != null ) {
+      if ( order.getClass().toString().equals(prop_.toString()) ) order = null;
```
Add braces to if it's matched with an else-if. The formatting of this is odd.

kgrgreer commented on src/foam/dao/index/TreeIndex.java:
```diff
@@ -74,6 +86,35 @@ public FindPlan planFind(Object state, Object key) {
+    }
+    ScanPlan selectPlan = new ScanPlan();
+    selectPlan.setCost(cost);
+    if ( predicate == null )
```
Why is this if statement needed?

kgrgreer commented on src/foam/dao/index/TreeIndex.java:
```diff
@@ -84,8 +125,20 @@ public SelectPlan planSelect(Object state, Sink sink, long skip, long limit, Com
+    if ( expr != null ) {
+      subTree = ((TreeNode) state).get((TreeNode) state, expr.getArg2().f(expr), prop_);
+      if ( subTree == null ) return new NotFoundPlan();
+      if ( ! (subTree instanceof TreeNode) )
```
Why is this check necessary? Why do you ever need to know what kind of subTree it is? You can't know all of the possibilities.

kgrgreer commented on src/foam/dao/index/TreeNode.java:
```diff
@@ -357,18 +361,62 @@ public TreeNode lte(TreeNode s, Object key, PropertyInfo prop) {
+    }
+    Object value = currentNode.getValue();
+    if ( value != null )
+      if ( value instanceof TreeNode ) selectAllHelper((TreeNode) value, sink);
```
Why is this check needed?

kgrgreer commented on src/foam/dao/index/TreeNode.java:
```diff
@@ -357,18 +361,62 @@ public TreeNode lte(TreeNode s, Object key, PropertyInfo prop) {
+      }
+    }
+    Object value = currentNode.getValue();
+    if ( value instanceof TreeNode ) {
```
Why needed?

kgrgreer commented on src/foam/dao/index/ValuePlan.java:
```diff
@@ -25,6 +24,12 @@ public FObject find(Object state, Object key) {
   public void select(Object state, Sink sink, long skip, long limit, Comparator order, Predicate predicate) {
-    sink.put((FObject) state, null);
+    if ( skip >= 1 || limit < 1 ) return;
+    if ( state instanceof TreeNode )
```
How can this happen? How can a Value state be a TreeNode?

amymeng5277 committed:
Merge pull request #16 from foam-framework/master

sync from master

amymeng5277 committed:
1. Eliminate cast type in TreeNode and ValuePlan. 2. Move calculate cost() function into ScanPlan. 3. Skip(),limit(),GT(),GTE().LT(),LTE() worked well now.

kgrgreer commented on src/foam/dao/index/ScanPlan.java:
```diff
@@ -6,19 +6,59 @@
+    boolean sortRequired = false;
+    boolean reverseSort = false;
+    if ( order_ != null ) {
+      if ( order_.getClass().toString().equals(propertyInfo.toString()) ) order_ = null;
```
Nest body of 'if' and add {}'s.

kgrgreer commented on src/foam/dao/index/ScanPlan.java:
```diff
@@ -6,19 +6,59 @@
+  public long calculateCost(PropertyInfo propertyInfo){
+    long cost;
+    if ( state_ == null ) {
+      cost = 0;
```
Isn't it faster/simpler to just return 0; here?

kgrgreer commented on src/foam/dao/index/ScanPlan.java:
```diff
@@ -6,19 +6,59 @@
+        reverseSort = true;
+      } else {
+        sortRequired = true;
+        if ( cost != 0 ) cost *= Math.log(cost) / Math.log(2);
```
if check not required.

kgrgreer commented on src/foam/dao/index/ScanPlan.java:
```diff
@@ -6,19 +6,59 @@
+    if ( order_ != null ) {
+      if ( order_.getClass().toString().equals(propertyInfo.toString()) ) order_ = null;
+      else if ( order_ instanceof Desc ) {
+        reverseSort = true;
```
Reverse sort still needs to check if the property is correct.

kgrgreer commented on src/foam/dao/index/ScanPlan.java:
```diff
@@ -6,19 +6,59 @@
+      }
+    }
+    if ( ! sortRequired ) {
+      if ( skip_ != 0 ) cost -= skip_;
```
This could result in - cost.

kgrgreer commented on src/foam/dao/index/ScanPlan.java:
```diff
@@ -27,6 +67,6 @@ public FObject find(Object state, Object key) {
     if(state == null)
       return;
-    ((TreeNode)state).select((TreeNode)state, sink, skip, limit, order, predicate);
+    ((TreeNode)state).select((TreeNode)state_, sink, skip_, limit_, order_, predicate_,tail_);
```
Add space before tail_.

kgrgreer commented on src/foam/dao/index/TreeIndex.java:
```diff
@@ -134,49 +106,38 @@ public SelectPlan planSelect(Object state, Sink sink, long skip, long limit, Com
-
-    return SelectPlanHelper(subTree, sink, skip, limit, order, predicate);
-
+    return new ScanPlan(subTree,sink,skip,limit,order,predicate,prop_,tail_);
```
Add spaces after commas.

kgrgreer commented on src/foam/dao/index/TreeNode.java:
```diff
@@ -265,9 +265,9 @@ public Object get(TreeNode s, Object key, PropertyInfo prop) {
-
     if ( r == 0 ) {
-      return s.value;
+      long size = s.value instanceof TreeNode?((TreeNode) s.value).size:1;
```
fix spacing

kgrgreer commented on src/foam/dao/index/TreeNode.java:
```diff
@@ -265,9 +265,9 @@ public Object get(TreeNode s, Object key, PropertyInfo prop) {
     if ( r == 0 ) {
-      return s.value;
+      long size = s.value instanceof TreeNode?((TreeNode) s.value).size:1;
+      return new TreeNode(s.key,s.value,size,0,null,null);
```
spacing

kgrgreer commented on src/foam/dao/index/TreeNode.java:
```diff
@@ -361,25 +361,23 @@ public TreeNode lte(TreeNode s, Object key, PropertyInfo prop) {
   }
 
-  protected void selectAllHelper(TreeNode currentNode, Sink sink){
+  protected void selectAllHelper(TreeNode currentNode, Sink sink, long skip, long limit, long size, Index tail){
```
rename to select_()

amymeng5277 committed:
1.Fixed space in TreeNode and ScanPlan 2. Change some part of costCalculate() function in Scanplan 3.Change name from selectAllHelper to select_

kgrgreer commented on src/foam/dao/index/ScanPlan.java:
```diff
@@ -36,24 +36,25 @@ public ScanPlan(Object state, Sink sink, long skip, long limit, Comparator order
     if ( state_ == null ) {
-      cost = 0;
+      return 0;
     } else {
```
No need for an 'else' if you've already returned.

amymeng5277 committed:
1. Fix 'else' in ScanPlan

amymeng5277 committed:
Merge pull request #17 from foam-framework/master

sync from master

amymeng5277 committed:
1.Add reverseSkipLimitTreeNode() function in TreeNode to implement Descending order. 2. Change Descending order's cost calculate.

amymeng5277 committed:
Merge pull request #18 from foam-framework/master

sync from master

amymeng5277 committed:
1.Add AND logic for MDAO 2.Changed isExMatch() function to simplifyPredicate() function 3. Add partialEval() function for And class.

amymeng5277 commented:
@kgrgreer Please review

kgrgreer commented on src/foam/mlang/mlangJava.js:
```diff
@@ -281,6 +281,44 @@ for ( int i = 0 ; i < length ; i++ ) {
+      javaReturns: 'foam.mlang.predicate.Predicate',
+      javaCode: 'java.util.List<Predicate> args = new java.util.ArrayList<>();\n' +
+      '    boolean update = false;\n' +
+      '    True TRUE = new True();\n' +
```
True and False should be Singletons.

amymeng5277 replied:
Ok, I will fix that.
